### PR TITLE
feat: support cascading reply

### LINKED ISSUE #333: Support reply cascading
The replies can have different layers, like this: https://casbin.org/docs/en/overview

We need to make sure:

1. In DB side, to keep consistency, every reply can have a parent reply. There will be a column in Reply DB table to store the parent reply ID. If the ID is empty, then this is a top-level reply, which has no parent reply (aka directly belonged to the topic). We DON'T store the child reply ID array in the Reply table, because adding or removing child replies is inconvenient.
2. We need to build the reply tree either in frontend or backend. I think backend would be better considering pagination. The Reply struct can have a []*Reply member to build the tree and it will be easy to serialize into JSON. But this member data is NOT stored in DB and will be populated at run-time in Go when fetching replies.

image

### PR DESCRIPTION
**Pull request by WindSpiritSR**
Deployed at: https://forum.leviatan.cn/
user: `testuser1@leviatan.cn`
password: `123`

Fix: https://github.com/casbin/casnode/issues/333

Signed-off-by: WindSpiritSR <simon343riley@gmail.com>

### CODE REVIEW

**Comment by hsluoyz** [comment]
@kocoler @Kininaru plz review

File: controllers/reply.go [hunk 1]
```diff
@@ -112,37 +112,27 @@ func (c *ApiController) AddReply() {
...
-	err = json.Unmarshal(c.Ctx.Input.RequestBody, &reply)
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &reply)
 	if err != nil {
 		panic(err)
 	}
 
-	balance := object.GetMemberBalance(memberId)
-	if balance < object.CreateReplyCost {
-		resp := Response{Status: "fail", Msg: "You don't have enough balance."}
+	if object.ContainsSensitiveWord(reply.Content) {
```
Comment (oranges-eating) [inline controllers/reply.go:134]: reply.content is ""?

File: object/reply.go [hunk 2]
```diff
@@ -65,6 +71,105 @@ func GetReplies(topicId int, memberId string, limit int, offset int) []*ReplyWit
...
+		} else {
+			if makeReplyTree(r.Child, reply) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getReplyLen(reply *ReplyWithAvatar) int {
```
Comment (kocoler) [inline object/reply.go:148]: Maybe use a map to optimize this recursion function is better.
This makes too many duplicated calculate.

File: web/src/main/ReplyBox.js [hunk 3]
```diff
@@ -299,6 +311,173 @@ class ReplyBox extends React.Component {
...
+            ) : null}
+            {this.props.account !== undefined && this.props.account !== null ? (
+              <a
+                href="#;"
+                onClick={() => {
+                  this.handleClick(`@${reply.author} `);
+                  this.setState({
+                    parent: {
+                      id:
+                        depth < Conf.ReplyMaxDepth - 1
```
Comment (kocoler) [inline web/src/main/ReplyBox.js]: The parent id shouldn't be affected by the `Conf` Instead, you could select the different style depending on the `depth` instead.
Our database should just record the truth.

**Review by kocoler — COMMENTED** [review]
The mobile style is broken. And the anchor of the nested reply is broken as well.

File: web/src/main/ReplyBox.js (hunk 3 above)
Comment (WindSpiritSR) [reply web/src/main/ReplyBox.js]: fixed

**Comment by WindSpiritSR** [comment]
@kocoler Mobile style fix by add `ReplyMobileMaxDepth` config, anchor is fixed

**Comment by hsluoyz** [comment]
@WindSpiritSR fix:

image

**Comment by WindSpiritSR** [comment]
@hsluoyz fixed

**Review by kocoler — APPROVED** [review]

**Pull request merged**